### PR TITLE
New RSX Debugger

### PR DESCRIPTION
**Pull request by elad335**
unknown

## Features
* RSX finally supports pausing and single-stepping on instructions!
* RSX command detail shows all arguments of a command, which was lacking from old rsx debugger. Press key I on command with a non-zero count. count is the number between the parenthesses.

### CODE REVIEW

File: rpcs3/rpcs3qt/breakpoint_list.h [hunk 1]
```diff
@@ -33,6 +33,6 @@ private Q_SLOTS:
 private:
 	breakpoint_handler* m_breakpoint_handler;
 
-	std::weak_ptr<cpu_thread> cpu;
-	std::shared_ptr<CPUDisAsm> m_disasm;
+	cpu_thread* cpu;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/breakpoint_list.h:36]:
```suggestion
	cpu_thread* m_cpu;
```

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 2]
```diff
@@ -503,20 +514,30 @@ void debugger_frame::UpdateUnitList()
...
+		if (!render && !weak.owner_before(cpu) && !cpu.owner_before(weak))
+		{
+			// They match, nothing to do.
+			return;
+		}
 
-	if (render && render == this->get_rsx())
-	{
-		return;
+		if (render && render == this->get_cpu())
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:537]: Why this-> ? Why not keep it consistent?

File: rpcs3/rpcs3qt/debugger_frame.h [hunk 3]
```diff
@@ -49,12 +51,13 @@ class debugger_frame : public custom_dock_widget
...
 	u64 m_threads_deleted = -1;
+	system_state m_emu_state{};
 	u32 m_last_pc = -1;
 	std::vector<char> m_last_query_state;
 	u32 m_last_step_over_breakpoint = -1;
 
-	std::shared_ptr<CPUDisAsm> m_disasm;
-	std::weak_ptr<cpu_thread> cpu;
+	std::shared_ptr<CPUDisAsm> m_disasm; // Only shared to allow base/derived functionality
+	std::shared_ptr<cpu_thread> cpu;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.h:60]: Might be a good time to change this to m_cpu

File: rpcs3/rpcs3qt/debugger_list.cpp [hunk 4]
```diff
@@ -123,13 +131,36 @@ void debugger_list::ShowAddress(u32 addr, bool force)
...
 	}
 
 	setLineWidth(-1);
 }
 
+void debugger_list::scroll(s32 steps)
+{
+	u32 result = m_pc;
+
+	//if (cpu && cpu->id_type() == 0x55)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_list.cpp:152]: ?

File: rpcs3/rpcs3qt/debugger_list.cpp [hunk 5]
```diff
@@ -139,21 +170,69 @@ void debugger_list::keyPressEvent(QKeyEvent* event)
...
+		m_cmd_detail->setFixedSize(m_cmd_detail->sizeHint());
+		m_cmd_detail->move(QCursor::pos());
+		return;
+	}
+
+	m_cmd_detail = new QDialog(this);
+	m_cmd_detail->setWindowTitle(tr("RSX Command Detail"));
+
+	m_detail_label = new QLabel(QString::fromStdString(rsx_dis.last_opcode), this);
+	m_detail_label->setFont([](QFont f) { f.setPointSize(9); return f; }(m_detail_label->font()));
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_list.cpp:207]: Maybe create a gui util function for the font. I'Ve seen you duplicating this code

And maybe consider mono font here

File: rpcs3/rpcs3qt/debugger_list.h [hunk 6]
```diff
@@ -44,6 +47,8 @@ public Q_SLOTS:
 	std::shared_ptr<gui_settings> xgui_settings;
 
 	breakpoint_handler* m_breakpoint_handler;
-	std::weak_ptr<cpu_thread> cpu;
-	std::shared_ptr<CPUDisAsm> m_disasm;
+	cpu_thread* cpu = nullptr;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_list.h:50]: m_cpu

File: rpcs3/rpcs3qt/instruction_editor_dialog.h [hunk 7]
```diff
@@ -23,7 +23,7 @@ class instruction_editor_dialog : public QDialog
 	QLabel* m_preview;
 
 public:
-	std::weak_ptr<cpu_thread> cpu;
+	std::shared_ptr<cpu_thread> cpu;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/instruction_editor_dialog.h:26]: m_cpu

File: rpcs3/rpcs3qt/register_editor_dialog.h [hunk 8]
```diff
@@ -20,7 +20,7 @@ class register_editor_dialog : public QDialog
 	QLineEdit* m_value_line;
 
 public:
-	std::weak_ptr<cpu_thread> cpu;
+	std::shared_ptr<cpu_thread> cpu;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/register_editor_dialog.h:23]: m_cpu

File: rpcs3/Emu/CPU/CPUDisAsm.h [hunk 9]
```diff
@@ -43,16 +47,19 @@ class CPUDisAsm
...
 				last_opcode = value + '\n';
 				break;
 			}
-			case CPUDisAsm_NormalMode:
+			case CPUDisAsmMode::Normal:
 			{
 				last_opcode = value;
 				break;
 			}
+			//case CPUDisAsmMode::List:
```
Comment (Megamouse) [inline rpcs3/Emu/CPU/CPUDisAsm.h:60]: ?

File: rpcs3/rpcs3qt/debugger_list.cpp (hunk 5 above)
Comment (elad335) [reply rpcs3/rpcs3qt/debugger_list.cpp:207]: Realized I can copy the font of debugger_list.

File: rpcs3/rpcs3qt/debugger_frame.h (hunk 3 above)
Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.h:60]: Done.

File: rpcs3/rpcs3qt/debugger_frame.cpp (hunk 2 above)
Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp:537]: Removed this->

File: rpcs3/Emu/CPU/CPUDisAsm.h (hunk 9 above)
Comment (elad335) [reply rpcs3/Emu/CPU/CPUDisAsm.h:60]: Removed.

File: rpcs3/rpcs3qt/register_editor_dialog.cpp [hunk 10]
```diff
@@ -144,7 +144,7 @@ register_editor_dialog::register_editor_dialog(QWidget *parent, const std::share
 	setModal(true);
 
 	// Events
-	connect(button_ok, &QAbstractButton::clicked, this, [=, this](){OnOkay(_cpu); accept();});
+	connect(button_ok, &QAbstractButton::clicked, this, [=, this](){ OnOkay(); accept(); });
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/register_editor_dialog.cpp:147]:
```suggestion
	connect(button_ok, &QAbstractButton::clicked, this, [this](){ OnOkay(); accept(); });
```

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 11]
```diff
@@ -291,8 +292,11 @@ void debugger_frame::keyPressEvent(QKeyEvent* event)
 		{
 		case Qt::Key_E:
 		{
-			instruction_editor_dialog* dlg = new instruction_editor_dialog(this, pc, cpu, m_disasm.get());
-			dlg->show();
+			if (m_cpu)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:295]: this changes the cpu object to m_cpu. was this intended or is this copy pasta from renaming cpu to m_cpu?

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 12]
```diff
@@ -302,13 +306,16 @@ void debugger_frame::keyPressEvent(QKeyEvent* event)
...
 
-			static_cast<spu_thread*>(cpu.get())->debugger_float_mode ^= 1; // Switch mode
+			static_cast<spu_thread*>(cpu)->debugger_float_mode ^= 1; // Switch mode
 			return;
 		}
 		case Qt::Key_R:
 		{
-			register_editor_dialog* dlg = new register_editor_dialog(this, cpu, m_disasm.get());
-			dlg->show();
+			if (m_cpu)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:314]: this changes the cpu object to m_cpu. was this intended or is this copy pasta from renaming cpu to m_cpu?

File: rpcs3/rpcs3qt/debugger_frame.cpp [hunk 13]
```diff
@@ -358,7 +365,7 @@ void debugger_frame::keyPressEvent(QKeyEvent* event)
 		case Qt::Key_M:
 		{
 			// Memory viewer
-			idm::make<memory_viewer_handle>(this, pc, cpu);
+			if (m_cpu) idm::make<memory_viewer_handle>(this, pc, m_cpu);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_frame.cpp:368]: this changes the cpu object to m_cpu. was this intended or is this copy pasta from renaming cpu to m_cpu?

File: rpcs3/rpcs3qt/debugger_frame.cpp (hunk 11 above)
Comment (elad335) [reply rpcs3/rpcs3qt/debugger_frame.cpp:295]: Intended.

File: rpcs3/rpcs3qt/debugger_list.cpp [hunk 14]
```diff
@@ -139,14 +168,84 @@ void debugger_list::keyPressEvent(QKeyEvent* event)
 
 	switch (event->key())
 	{
-	case Qt::Key_PageUp:   ShowAddress(m_pc - (m_item_count * 4), true); return;
-	case Qt::Key_PageDown: ShowAddress(m_pc + (m_item_count * 4), true); return;
-	case Qt::Key_Up:       ShowAddress(m_pc - 4, true); return;
-	case Qt::Key_Down:     ShowAddress(m_pc + 4, true); return;
+	case Qt::Key_PageUp:   scroll(0 - m_item_count); return;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/debugger_list.cpp:171]: did you add these to the help page?

File: rpcs3/Emu/CPU/CPUDisAsm.h [hunk 15]
```diff
@@ -4,26 +4,30 @@
 #include <limits>
 #include "Utilities/StrFmt.h"
 
-enum CPUDisAsmMode
+enum class CPUDisAsmMode
```
Comment (Nekotekina) [inline rpcs3/Emu/CPU/CPUDisAsm.h:7]: maybe `cpu_disasm_mode`

**Review by Nekotekina — CHANGES_REQUESTED** [review]

File: rpcs3/Emu/CPU/CPUDisAsm.h (hunk 15 above)
Comment (elad335) [reply rpcs3/Emu/CPU/CPUDisAsm.h]: Fixed.

**Comment by Yahfz** [comment]
@elad335 
This build fails to compile ppu modules after firmware installation
image

**Pull request merged**